src/sha256.nr

```diff
@@ -63,94 +29,132 @@ pub fn digest<let N: u32>(msg: [u8; N]) -> HASH {
+        let modulo = N % BLOCK_SIZE;
+        // Handle setup of the final msg block.
+        // This case is only hit if the msg is less than the block size,
+        // or our message cannot be evenly split into blocks.
+        if modulo != 0 {
+            let msg_start = BLOCK_SIZE * num_blocks;
+            let (new_msg_block, new_msg_byte_ptr) =
+                unsafe { build_msg_block(msg, message_size, msg_start) };
 
-        if msg_start < message_size {
-            msg_block = new_msg_block;
-        }
+            if msg_start < message_size {
+                msg_block = new_msg_block;
+            }
 
-        if !is_unconstrained() {
             let new_msg_byte_ptr = verify_msg_block(msg, message_size, msg_block, msg_start);
             if msg_start < message_size {
                 msg_byte_ptr = new_msg_byte_ptr;
                 verify_msg_block_padding(msg_block, msg_byte_ptr);
             }
-        } else if msg_start < message_size {
-            msg_byte_ptr = new_msg_byte_ptr;
         }
+
+        // If we had modulo == 0 then it means the last block was full,
+        // and we can reset the pointer to zero to overwrite it.
+        if msg_byte_ptr == BLOCK_SIZE {
+            msg_byte_ptr = 0;
+        }
+
+        // Pad the rest such that we have a [u32; 2] block at the end representing the length
+        // of the message, and a block of 1 0 ... 0 following the message (i.e. [1 << 7, 0, ..., 0]).
+        // Here we rely on the fact that everything beyond the available input is set to 0.
+        let index = msg_byte_ptr / INT_SIZE;
+        msg_block[index] = set_item_byte_then_zeros(msg_block[index], msg_byte_ptr, 1 << 7);
+
+        msg_byte_ptr = msg_byte_ptr + 1;
+        let last_block = msg_block;
+
+        // If we don't have room to write the size, compress the block and reset it.
+        if msg_byte_ptr > MSG_SIZE_PTR {
+            h = sha256_compression(msg_block, h);
+            // `attach_len_to_msg_block` will zero out everything after the `msg_byte_ptr`.
+            msg_byte_ptr = 0;
+        }
+
+        msg_block = unsafe { attach_len_to_msg_block(msg_block, msg_byte_ptr, message_size) };
+
+        verify_msg_len(msg_block, last_block, msg_byte_ptr, message_size);
+
+        hash_final_block(msg_block, h)
     }
+}
 
-    // If we had modulo == 0 then it means the last block was full,
-    // and we can reset the pointer to zero to overwrite it.
-    if msg_byte_ptr == BLOCK_SIZE {
-        msg_byte_ptr = 0;
+// Variable size SHA-256 hash
+unconstrained fn __sha256_var<let N: u32>(msg: [u8; N], message_size: u32) -> HASH {
+    let num_full_blocks = message_size / BLOCK_SIZE;
+    // Intermediate hash, starting with the canonical initial value
+    let mut h: STATE = INITIAL_STATE;
+    // Pointer into msg_block on a 64 byte scale
+    for i in 0..num_full_blocks {
+        let (msg_block, _) = build_msg_block(msg, message_size, BLOCK_SIZE * i);
```

Comment:
As we don't use the `msg_byte_ptr` value, we can check to make sure that we're not unnecessarily calculating this value in `build_msg_block` for this case.